File: test_wasm_advanced.lua
-- wasm3 C语言高级功能测试
-- 运行: lxclua test_wasm_advanced.lua

local wasm3 = require("wasm3")

print("=== wasm3 C语言高级功能测试 ===\n")

-- 创建环境
local env = wasm3.newEnvironment()
local runtime = env:newRuntime(2 * 1024 * 1024)  -- 2MB 栈

-- 读取 WASM 文件
local f = io.open("test_wasm_advanced.wasm", "rb")
if not f then
    print("[错误] 请先编译: make wasm-c-all SRC=test_wasm_advanced.c")
    return
end
local wasm_data = f:read("*a")
f:close()

-- 解析模块
local module = env:parseModule(wasm_data)

-- 加载模块
runtime:loadModule(module)

-- 辅助函数
local function call(name, ...)
    local func = runtime:findFunction(name)
    return func:call(...)
end

local function test(name, got, expected)
    local ok = got == expected
    local status = ok and "[OK]" or "[FAIL]"
    print(string.format("%s %s: got=%s, expected=%s", status, name, tostring(got), tostring(expected)))
    return ok
end

local all_passed = true

-- ==================== 数学运算 ====================
print("--- 数学运算 ---")

all_passed = test("power(2, 10)", call("power", 2, 10), 1024) and all_passed
all_passed = test("power(3, 5)", call("power", 3, 5), 243) and all_passed
all_passed = test("gcd(48, 18)", call("gcd", 48, 18), 6) and all_passed
all_passed = test("gcd(100, 35)", call("gcd", 100, 35), 5) and all_passed
all_passed = test("lcm(12, 18)", call("lcm", 12, 18), 36) and all_passed
all_passed = test("is_prime(17)", call("is_prime", 17), 1) and all_passed
all_passed = test("is_prime(18)", call("is_prime", 18), 0) and all_passed
all_passed = test("nth_prime(10)", call("nth_prime", 10), 29) and all_passed
print()

-- ==================== 位操作 ====================
print("--- 位操作 ---")

all_passed = test("popcount(255)", call("popcount", 255), 8) and all_passed
all_passed = test("popcount(0xAA)", call("popcount", 0xAA), 4) and all_passed
all_passed = test("rotl(1, 4)", call("rotl", 1, 4), 16) and all_passed
all_passed = test("rotr(16, 4)", call("rotr", 16, 4), 1) and all_passed
print()

-- ==================== 递归算法 ====================
print("--- 递归算法 ---")

all_passed = test("hanoi_moves(3)", call("hanoi_moves", 3), 7) and all_passed
all_passed = test("hanoi_moves(10)", call("hanoi_moves", 10), 1023) and all_passed
all_passed = test("combination(5, 2)", call("combination", 5, 2), 10) and all_passed
all_passed = test("combination(10, 5)", call("combination", 10, 5), 252) and all_passed
all_passed = test("pascal(6, 3)", call("pascal", 6, 3), 20) and all_passed
print()

-- ==================== 数列 ====================
print("--- 数列 ---")

all_passed = test("fib(10)", call("fib", 10), 55) and all_passed
all_passed = test("fib(20)", call("fib", 20), 6765) and all_passed
all_passed = test("lucas(10)", call("lucas", 10), 123) and all_passed
all_passed = test("pell(10)", call("pell", 10), 2378) and all_passed
all_passed = test("triangular(10)", call("triangular", 10), 55) and all_passed
all_passed = test("pentagonal(10)", call("pentagonal", 10), 145) and all_passed
all_passed = test("hexagonal(10)", call("hexagonal", 10), 190) and all_passed
print()

-- ==================== 结构体操作 ====================
print("--- 结构体操作 ---")

-- Point 结构体测试 - 使用分开的坐标参数函数
local dist_sq = call("point_distance_sq_xy", 3, 4, 0, 0)
all_passed = test("point_distance_sq(3,4)-(0,0)", dist_sq, 25) and all_passed

local manhattan = call("point_manhattan_xy", 3, 4, 0, 0)
all_passed = test("point_manhattan(3,4)-(0,0)", manhattan, 7) and all_passed

local dist = call("point_distance_xy", 3, 4, 0, 0)
print(string.format("[OK] point_distance(3,4)-(0,0) = %.2f", dist))

-- Rectangle 结构体测试
local area = call("rect_area_wh", 10, 20)
all_passed = test("rect_area(10x20)", area, 200) and all_passed

local perimeter = call("rect_perimeter_wh", 10, 20)
all_passed = test("rect_perimeter(10x20)", perimeter, 60) and all_passed

local contains = call("rect_contains_point_wh", 10, 20, 5, 5)
all_passed = test("rect_contains(5,5)", contains, 1) and all_passed

local not_contains = call("rect_contains_point_wh", 10, 20, 15, 5)
all_passed = test("rect_contains(15,5)", not_contains, 0) and all_passed
print()

-- ==================== 内存操作 ====================
print("--- 内存操作 ---")

local mem_size = runtime:getMemorySize()
print(string.format("[INFO] 内存大小: %d bytes", mem_size))

-- 使用 WASM 函数读写内存
local test_offset = 1024

-- 写入并读取 int32
call("write_i32", test_offset, 0, 12345)
local val = call("read_i32", test_offset, 0)
all_passed = test("memory write/read i32", val, 12345) and all_passed

-- 写入并读取 int64
call("write_i64", test_offset, 0, 0x123456789ABCDEF0)
local val64 = call("read_i64", test_offset, 0)
print(string.format("[OK] memory write/read i64: 0x%X", val64))

-- 写入并读取 float
call("write_f32", test_offset, 0, 3.14159)
local f32 = call("read_f32", test_offset, 0)
print(string.format("[OK] memory write/read f32: %.5f", f32))

-- 写入并读取 double
call("write_f64", test_offset, 0, 2.718281828)
local f64 = call("read_f64", test_offset, 0)
print(string.format("[OK] memory write/read f64: %.9f", f64))
print()

-- ==================== 数组操作 ====================
print("--- 数组操作 ---")

-- 在内存中创建数组
local arr_offset = 2048
local arr_len = 5
local arr_data = {10, 20, 30, 40, 50}

-- 写入数组
for i, v in ipairs(arr_data) do
    call("write_i32", arr_offset, i - 1, v)
end

-- 测试数组函数
local sum = call("array_sum", arr_offset, arr_len)
all_passed = test("array_sum({10,20,30,40,50})", sum, 150) and all_passed

local max = call("array_max", arr_offset, arr_len)
all_passed = test("array_max({10,20,30,40,50})", max, 50) and all_passed

local min = call("array_min", arr_offset, arr_len)
all_passed = test("array_min({10,20,30,40,50})", min, 10) and all_passed

-- 数组反转
call("array_reverse", arr_offset, arr_len)
print("[OK] 数组反转后:")
for i = 0, arr_len - 1 do
    io.write(string.format("  arr[%d]=%d", i, call("read_i32", arr_offset, i)))
end
print()

-- 恢复数组
for i, v in ipairs(arr_data) do
    call("write_i32", arr_offset, i - 1, v)
end

-- 快速排序测试
local unsorted_offset = 2100
local unsorted = {64, 34, 25, 12, 22, 11, 90}
for i, v in ipairs(unsorted) do
    call("write_i32", unsorted_offset, i - 1, v)
end
call("array_sort_quick", unsorted_offset, 7)
print("[OK] 快速排序后:")
for i = 0, 6 do
    io.write(string.format("%d ", call("read_i32", unsorted_offset, i)))
end
print()

-- 二分查找
local found = call("array_binary_search", unsorted_offset, 7, 25)
all_passed = test("binary_search(25)", found, 3) and all_passed  -- 排序后 25 在索引 3
print()

-- ==================== 矩阵操作 ====================
print("--- 矩阵操作 ---")

-- 创建 3x3 矩阵
local mat_offset = 4096
for i = 0, 8 do
    call("write_i32", mat_offset, i, i + 1)
end

print("[OK] 3x3 矩阵:")
for row = 0, 2 do
    io.write("  [")
    for col = 0, 2 do
        local val = call("matrix_get", mat_offset, 3, row, col)
        io.write(string.format("%2d", val))
        if col < 2 then io.write(", ") end
    end
    io.write("]\n")
end

local mat_sum = call("matrix_sum", mat_offset, 3, 3)
all_passed = test("matrix_sum", mat_sum, 45) and all_passed

local trace = call("matrix_trace", mat_offset, 3)
all_passed = test("matrix_trace", trace, 15) and all_passed

local det = call("matrix_det_3x3", mat_offset)
print(string.format("[OK] 行列式 det = %d", det))
print()

-- ==================== 高级算法 ====================
print("--- 高级算法 ---")

-- 最长递增子序列
local lis_offset = 2300
local lis_data = {10, 22, 9, 33, 21, 50, 41}
for i, v in ipairs(lis_data) do
    call("write_i32", lis_offset, i - 1, v)
end
local lis = call("lis_length", lis_offset, 7)
all_passed = test("lis_length({10,22,9,33,21,50,41})", lis, 4) and all_passed  -- 正确答案是 4

-- 最大子数组和
local msa_offset = 2400
local msa_data = {-2, 1, -3, 4, -1, 2, 1, -5, 4}
for i, v in ipairs(msa_data) do
    call("write_i32", msa_offset, i - 1, v)
end
local msa = call("max_subarray_sum", msa_offset, 9)
all_passed = test("max_subarray_sum", msa, 6) and all_passed
print()

-- ==================== 结果 ====================
print("\n" .. string.rep("=", 50))
if all_passed then
    print("=== 所有测试通过! ===")
else
    print("=== 部分测试失败! ===")
end
print(string.rep("=", 50))

File: test_wasm_advanced.c
/*
 * wasm3 C语言高级功能测试模块（纯计算，无系统调用）
 * 编译: make wasm-c-all SRC=test_wasm_advanced.c
 */

#include <stdint.h>

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#define EXPORT EMSCRIPTEN_KEEPALIVE
#else
#define EXPORT
#endif

/* ==================== 结构体定义 ==================== */

typedef struct {
    int32_t x;
    int32_t y;
} Point;

typedef struct {
    int32_t width;
    int32_t height;
} Rectangle;

/* ==================== 内存操作（使用线性内存） ==================== */

/* 读取 int32 */
EXPORT int32_t read_i32(const void* ptr, int32_t offset) {
    const int32_t* p = (const int32_t*)ptr;
    return p[offset];
}

/* 写入 int32 */
EXPORT void write_i32(void* ptr, int32_t offset, int32_t value) {
    int32_t* p = (int32_t*)ptr;
    p[offset] = value;
}

/* 读取 int64 */
EXPORT int64_t read_i64(const void* ptr, int32_t offset) {
    const int64_t* p = (const int64_t*)ptr;
    return p[offset];
}

/* 写入 int64 */
EXPORT void write_i64(void* ptr, int32_t offset, int64_t value) {
    int64_t* p = (int64_t*)ptr;
    p[offset] = value;
}

/* 读取 float */
EXPORT float read_f32(const void* ptr, int32_t offset) {
    const float* p = (const float*)ptr;
    return p[offset];
}

/* 写入 float */
EXPORT void write_f32(void* ptr, int32_t offset, float value) {
    float* p = (float*)ptr;
    p[offset] = value;
}

/* 读取 double */
EXPORT double read_f64(const void* ptr, int32_t offset) {
    const double* p = (const double*)ptr;
    return p[offset];
}

/* 写入 double */
EXPORT void write_f64(void* ptr, int32_t offset, double value) {
    double* p = (double*)ptr;
    p[offset] = value;
}

/* ==================== 数组操作 ==================== */

/* 数组求和 */
EXPORT int32_t array_sum(const int32_t* arr, int32_t len) {
    int32_t sum = 0;
    for (int32_t i = 0; i < len; i++) {
        sum += arr[i];
    }
    return sum;
}

/* 数组最大值 */
EXPORT int32_t array_max(const int32_t* arr, int32_t len) {
    if (len <= 0) return 0;
    int32_t max = arr[0];
    for (int32_t i = 1; i < len; i++) {
        if (arr[i] > max) max = arr[i];
    }
    return max;
}

/* 数组最小值 */
EXPORT int32_t array_min(const int32_t* arr, int32_t len) {
    if (len <= 0) return 0;
    int32_t min = arr[0];
    for (int32_t i = 1; i < len; i++) {
        if (arr[i] < min) min = arr[i];
    }
    return min;
}

/* 数组平均值 */
EXPORT double array_avg(const int32_t* arr, int32_t len) {
    if (len <= 0) return 0.0;
    int64_t sum = 0;
    for (int32_t i = 0; i < len; i++) {
        sum += arr[i];
    }
    return (double)sum / len;
}

/* 数组反转 */
EXPORT void array_reverse(int32_t* arr, int32_t len) {
    for (int32_t i = 0; i < len / 2; i++) {
        int32_t tmp = arr[i];
        arr[i] = arr[len - 1 - i];
        arr[len - 1 - i] = tmp;
    }
}

/* 数组排序（冒泡） */
EXPORT void array_sort_bubble(int32_t* arr, int32_t len) {
    for (int32_t i = 0; i < len - 1; i++) {
        for (int32_t j = 0; j < len - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                int32_t tmp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = tmp;
            }
        }
    }
}

/* 数组排序（快速排序） */
static void quick_sort_impl(int32_t* arr, int32_t low, int32_t high) {
    if (low < high) {
        int32_t pivot = arr[high];
        int32_t i = low - 1;
        for (int32_t j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                int32_t tmp = arr[i];
                arr[i] = arr[j];
                arr[j] = tmp;
            }
        }
        int32_t tmp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = tmp;
        int32_t pi = i + 1;
        quick_sort_impl(arr, low, pi - 1);
        quick_sort_impl(arr, pi + 1, high);
    }
}

EXPORT void array_sort_quick(int32_t* arr, int32_t len) {
    quick_sort_impl(arr, 0, len - 1);
}

/* 二分查找 */
EXPORT int32_t array_binary_search(const int32_t* arr, int32_t len, int32_t target) {
    int32_t left = 0, right = len - 1;
    while (left <= right) {
        int32_t mid = left + (right - left) / 2;
        if (arr[mid] == target) return mid;
        if (arr[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return -1;
}

/* 数组填充 */
EXPORT void array_fill(int32_t* arr, int32_t len, int32_t value) {
    for (int32_t i = 0; i < len; i++) {
        arr[i] = value;
    }
}

/* 数组拷贝 */
EXPORT void array_copy(int32_t* dst, const int32_t* src, int32_t len) {
    for (int32_t i = 0; i < len; i++) {
        dst[i] = src[i];
    }
}

/* 数组查找 */
EXPORT int32_t array_find(const int32_t* arr, int32_t len, int32_t value) {
    for (int32_t i = 0; i < len; i++) {
        if (arr[i] == value) return i;
    }
    return -1;
}

/* 数组计数 */
EXPORT int32_t array_count(const int32_t* arr, int32_t len, int32_t value) {
    int32_t count = 0;
    for (int32_t i = 0; i < len; i++) {
        if (arr[i] == value) count++;
    }
    return count;
}

/* ==================== Point 结构体操作 ==================== */

/* 创建点（返回结构体值） */
EXPORT Point point_create(int32_t x, int32_t y) {
    Point p = {x, y};
    return p;
}

/* 获取 X 坐标 */
EXPORT int32_t point_get_x(Point p) {
    return p.x;
}

/* 获取 Y 坐标 */
EXPORT int32_t point_get_y(Point p) {
    return p.y;
}

/* 点距离平方（使用分开的坐标参数） */
EXPORT int64_t point_distance_sq_xy(int32_t x1, int32_t y1, int32_t x2, int32_t y2) {
    int64_t dx = x1 - x2;
    int64_t dy = y1 - y2;
    return dx * dx + dy * dy;
}

/* 点距离平方 */
EXPORT int64_t point_distance_sq(Point a, Point b) {
    return point_distance_sq_xy(a.x, a.y, b.x, b.y);
}

/* 点距离（浮点，使用分开的坐标参数） */
EXPORT double point_distance_xy(int32_t x1, int32_t y1, int32_t x2, int32_t y2) {
    int64_t sq = point_distance_sq_xy(x1, y1, x2, y2);
    return __builtin_sqrt((double)sq);
}

/* 点距离（浮点） */
EXPORT double point_distance(Point a, Point b) {
    return point_distance_xy(a.x, a.y, b.x, b.y);
}

/* 点相等判断 */
EXPORT int32_t point_equals(Point a, Point b) {
    return a.x == b.x && a.y == b.y;
}

/* 点的曼哈顿距离（使用分开的坐标参数） */
EXPORT int32_t point_manhattan_xy(int32_t x1, int32_t y1, int32_t x2, int32_t y2) {
    int32_t dx = x1 > x2 ? x1 - x2 : x2 - x1;
    int32_t dy = y1 > y2 ? y1 - y2 : y2 - y1;
    return dx + dy;
}

/* 点的曼哈顿距离 */
EXPORT int32_t point_manhattan(Point a, Point b) {
    return point_manhattan_xy(a.x, a.y, b.x, b.y);
}

/* 点移动 */
EXPORT Point point_move(Point p, int32_t dx, int32_t dy) {
    p.x += dx;
    p.y += dy;
    return p;
}

/* ==================== Rectangle 结构体操作 ==================== */

/* 创建矩形 */
EXPORT Rectangle rect_create(int32_t width, int32_t height) {
    Rectangle r = {width, height};
    return r;
}

/* 计算面积（使用分开的参数） */
EXPORT int32_t rect_area_wh(int32_t width, int32_t height) {
    return width * height;
}

/* 计算面积 */
EXPORT int32_t rect_area(Rectangle r) {
    return rect_area_wh(r.width, r.height);
}

/* 计算周长（使用分开的参数） */
EXPORT int32_t rect_perimeter_wh(int32_t width, int32_t height) {
    return 2 * (width + height);
}

/* 计算周长 */
EXPORT int32_t rect_perimeter(Rectangle r) {
    return rect_perimeter_wh(r.width, r.height);
}

/* 判断点是否在矩形内（使用分开的参数） */
EXPORT int32_t rect_contains_point_wh(int32_t width, int32_t height, int32_t x, int32_t y) {
    return x >= 0 && x < width && y >= 0 && y < height;
}

/* 判断点是否在矩形内 */
EXPORT int32_t rect_contains_point(Rectangle r, int32_t x, int32_t y) {
    return rect_contains_point_wh(r.width, r.height, x, y);
}

/* 矩形缩放 */
EXPORT Rectangle rect_scale(Rectangle r, int32_t factor) {
    r.width *= factor;
    r.height *= factor;
    return r;
}

/* ==================== 数学运算 ==================== */

/* 幂运算 */
EXPORT int32_t power(int32_t base, int32_t exp) {
    int32_t result = 1;
    while (exp > 0) {
        if (exp & 1) result *= base;
        base *= base;
        exp >>= 1;
    }
    return result;
}

/* 最大公约数 */
EXPORT int32_t gcd(int32_t a, int32_t b) {
    while (b != 0) {
        int32_t tmp = b;
        b = a % b;
        a = tmp;
    }
    return a;
}

/* 最小公倍数 */
EXPORT int32_t lcm(int32_t a, int32_t b) {
    return (a / gcd(a, b)) * b;
}

/* 判断素数 */
EXPORT int32_t is_prime(int32_t n) {
    if (n <= 1) return 0;
    if (n <= 3) return 1;
    if (n % 2 == 0 || n % 3 == 0) return 0;
    for (int32_t i = 5; i * i <= n; i += 6) {
        if (n % i == 0 || n % (i + 2) == 0) return 0;
    }
    return 1;
}

/* 第n个素数 */
EXPORT int32_t nth_prime(int32_t n) {
    int32_t count = 0;
    int32_t num = 2;
    while (count < n) {
        if (is_prime(num)) count++;
        if (count < n) num++;
    }
    return num;
}

/* ==================== 位操作 ==================== */

/* 位计数 */
EXPORT int32_t popcount(uint32_t x) {
    int32_t count = 0;
    while (x) {
        count += x & 1;
        x >>= 1;
    }
    return count;
}

/* 反转位 */
EXPORT uint32_t reverse_bits(uint32_t x) {
    uint32_t result = 0;
    for (int i = 0; i < 32; i++) {
        result = (result << 1) | (x & 1);
        x >>= 1;
    }
    return result;
}

/* 循环左移 */
EXPORT uint32_t rotl(uint32_t x, int32_t n) {
    n &= 31;
    return (x << n) | (x >> (32 - n));
}

/* 循环右移 */
EXPORT uint32_t rotr(uint32_t x, int32_t n) {
    n &= 31;
    return (x >> n) | (x << (32 - n));
}

/* ==================== 递归算法 ==================== */

/* 汉诺塔移动次数 */
EXPORT int64_t hanoi_moves(int32_t n) {
    if (n <= 0) return 0;
    if (n == 1) return 1;
    return 2 * hanoi_moves(n - 1) + 1;
}

/* 组合数 C(n, k) */
EXPORT int64_t combination(int32_t n, int32_t k) {
    if (k < 0 || k > n) return 0;
    if (k == 0 || k == n) return 1;
    return combination(n - 1, k - 1) + combination(n - 1, k);
}

/* 杨辉三角第n行第k列 */
EXPORT int64_t pascal(int32_t n, int32_t k) {
    return combination(n, k);
}

/* ==================== 斐波那契变体 ==================== */

/* 斐波那契（迭代） */
EXPORT int64_t fib(int32_t n) {
    if (n <= 0) return 0;
    if (n == 1) return 1;
    int64_t a = 0, b = 1;
    for (int32_t i = 2; i <= n; i++) {
        int64_t tmp = a + b;
        a = b;
        b = tmp;
    }
    return b;
}

/* 卢卡斯数 */
EXPORT int64_t lucas(int32_t n) {
    if (n == 0) return 2;
    if (n == 1) return 1;
    int64_t a = 2, b = 1;
    for (int32_t i = 2; i <= n; i++) {
        int64_t tmp = a + b;
        a = b;
        b = tmp;
    }
    return b;
}

/* 佩尔数 */
EXPORT int64_t pell(int32_t n) {
    if (n == 0) return 0;
    if (n == 1) return 1;
    int64_t a = 0, b = 1;
    for (int32_t i = 2; i <= n; i++) {
        int64_t tmp = 2 * b + a;
        a = b;
        b = tmp;
    }
    return b;
}

/* ==================== 三角数 ==================== */

/* 三角数 */
EXPORT int64_t triangular(int32_t n) {
    return (int64_t)n * (n + 1) / 2;
}

/* 五角数 */
EXPORT int64_t pentagonal(int32_t n) {
    return (int64_t)n * (3 * n - 1) / 2;
}

/* 六角数 */
EXPORT int64_t hexagonal(int32_t n) {
    return (int64_t)n * (2 * n - 1);
}

/* ==================== 矩阵操作 ==================== */

/* 矩阵获取元素 */
EXPORT int32_t matrix_get(const int32_t* m, int32_t cols, int32_t row, int32_t col) {
    return m[row * cols + col];
}

/* 矩阵设置元素 */
EXPORT void matrix_set(int32_t* m, int32_t cols, int32_t row, int32_t col, int32_t value) {
    m[row * cols + col] = value;
}

/* 矩阵填充 */
EXPORT void matrix_fill(int32_t* m, int32_t rows, int32_t cols, int32_t value) {
    for (int32_t i = 0; i < rows * cols; i++) {
        m[i] = value;
    }
}

/* 矩阵求和 */
EXPORT int64_t matrix_sum(const int32_t* m, int32_t rows, int32_t cols) {
    int64_t sum = 0;
    for (int32_t i = 0; i < rows * cols; i++) {
        sum += m[i];
    }
    return sum;
}

/* 矩阵转置 */
EXPORT void matrix_transpose(const int32_t* src, int32_t* dst, int32_t rows, int32_t cols) {
    for (int32_t i = 0; i < rows; i++) {
        for (int32_t j = 0; j < cols; j++) {
            dst[j * rows + i] = src[i * cols + j];
        }
    }
}

/* 矩阵迹（对角线元素和） */
EXPORT int32_t matrix_trace(const int32_t* m, int32_t n) {
    int32_t trace = 0;
    for (int32_t i = 0; i < n; i++) {
        trace += m[i * n + i];
    }
    return trace;
}

/* 3x3 矩阵行列式 */
EXPORT int32_t matrix_det_3x3(const int32_t* m) {
    return m[0] * (m[4] * m[8] - m[5] * m[7])
         - m[1] * (m[3] * m[8] - m[5] * m[6])
         + m[2] * (m[3] * m[7] - m[4] * m[6]);
}

/* ==================== 回调函数模拟 ==================== */

/* 内置变换函数 */
EXPORT int32_t transform_double(int32_t x) { return x * 2; }
EXPORT int32_t transform_square(int32_t x) { return x * x; }
EXPORT int32_t transform_negate(int32_t x) { return -x; }
EXPORT int32_t transform_abs(int32_t x) { return x < 0 ? -x : x; }

/* 对数组应用变换 */
EXPORT void array_transform(int32_t* arr, int32_t len, int32_t transform_id) {
    int32_t (*funcs[4])(int32_t) = {
        transform_double,
        transform_square,
        transform_negate,
        transform_abs
    };
    int32_t (*f)(int32_t) = funcs[transform_id % 4];
    for (int32_t i = 0; i < len; i++) {
        arr[i] = f(arr[i]);
    }
}

/* ==================== 高级算法 ==================== */

/* 动态规划：最大子数组和 */
EXPORT int32_t max_subarray_sum(const int32_t* arr, int32_t len) {
    if (len <= 0) return 0;
    
    int32_t max_sum = arr[0];
    int32_t current_sum = arr[0];
    
    for (int32_t i = 1; i < len; i++) {
        if (current_sum > 0) {
            current_sum += arr[i];
        } else {
            current_sum = arr[i];
        }
        if (current_sum > max_sum) {
            max_sum = current_sum;
        }
    }
    return max_sum;
}

/* 动态规划：最长递增子序列长度（简化版，使用固定大小数组） */
#define LIS_MAX_LEN 100
static int32_t lis_dp[LIS_MAX_LEN];

EXPORT int32_t lis_length(const int32_t* arr, int32_t len) {
    if (len <= 0) return 0;
    if (len > LIS_MAX_LEN) len = LIS_MAX_LEN;
    
    for (int32_t i = 0; i < len; i++) {
        lis_dp[i] = 1;
        for (int32_t j = 0; j < i; j++) {
            if (arr[j] < arr[i] && lis_dp[j] + 1 > lis_dp[i]) {
                lis_dp[i] = lis_dp[j] + 1;
            }
        }
    }
    
    int32_t max_len = lis_dp[0];
    for (int32_t i = 1; i < len; i++) {
        if (lis_dp[i] > max_len) max_len = lis_dp[i];
    }
    
    return max_len;
}

/* ==================== 字符串操作 ==================== */

/* 字符串长度 */
EXPORT int32_t str_length(const char* s) {
    if (!s) return 0;
    int32_t len = 0;
    while (s[len]) len++;
    return len;
}

/* 字符串哈希 */
EXPORT uint32_t str_hash(const char* s) {
    uint32_t hash = 5381;
    while (*s) {
        hash = ((hash << 5) + hash) + (unsigned char)*s;
        s++;
    }
    return hash;
}

/* 字符串反转 */
EXPORT void str_reverse(char* s) {
    int32_t len = str_length(s);
    for (int32_t i = 0; i < len / 2; i++) {
        char tmp = s[i];
        s[i] = s[len - 1 - i];
        s[len - 1 - i] = tmp;
    }
}

/* 转大写 */
EXPORT void str_to_upper(char* s) {
    while (*s) {
        if (*s >= 'a' && *s <= 'z') {
            *s = *s - 'a' + 'A';
        }
        s++;
    }
}

/* 转小写 */
EXPORT void str_to_lower(char* s) {
    while (*s) {
        if (*s >= 'A' && *s <= 'Z') {
            *s = *s - 'A' + 'a';
        }
        s++;
    }
}
